Add http api for fetching accounts

laurb9 opened the pull request:
This works the same way as the API for stored requests,
preserving the ability to query multiple ids, so the
same API endpoint can be extended to support accounts.

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -81,8 +98,68 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
+	respBytes, err := ioutil.ReadAll(httpResp.Body)
+	if err != nil {
+		return nil, []error{
+			fmt.Errorf(`Error fetching accounts %v via http: could not read response`, accountIDs),
```
We should also append the error to this log message. I see you added that for the error messages above so looks like this was missed by mistake?

laurb9 replied:
Replacing with
```go
			fmt.Errorf(`Error fetching accounts %v via http: error reading response: %v`, accountIDs, err),
```

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -34,20 +39,32 @@ import (
 
-	glog.Info("Making http_fetcher which calls GET " + urlPrefix + "request-ids=%REQUEST_ID_LIST%&imp-ids=%IMP_ID_LIST%")
+	if url, err := url.Parse(endpoint); err != nil {
+		glog.Warningf(`Invalid endpoint "%s": %v`, endpoint, err)
```
Thanks for adding this. Although, please consider moving this before creating `urlPrefix` on L56 because if the `endpoint` isn't valid, there's not much point in appending `?` or `&` to it.

laurb9 replied:
Done. Should it be an error instead of warning ?

I intend to explore replacing Endpoint with URL also, perhaps in another PR.

mansinahar replied:
That's a good point. I'd say it should be a fatal error because if the endpoint isn't valid, the fetching wouldn't work. I see the DB fetcher also doing the same in case of invalid config. Ideally these things should be part of a validate method that validates the config for stored requests

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -81,8 +98,68 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
+	httpReq, err := http.NewRequestWithContext(ctx, "GET", fetcher.Endpoint+"account-ids=[\""+strings.Join(accountIDs, "\",\"")+"\"]", nil)
+	if err != nil {
+		return nil, []error{
+			fmt.Errorf(`Error fetching accounts "%v" via http: %v`, accountIDs, err),
```
Please consider changing this error message to `Error building the http request to fetch accounts "%v"` or something along those lines so that it is clear that the error was in building the request rather than fetching the account data with a valid http request

laurb9 replied:
Will change it to
```go
fmt.Errorf(`Error fetching accounts "%v" via http: build request failed with %v`, accountIDs, err),
```

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -81,8 +98,68 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
+	var responseData accountsResponseContract
+	if err = json.Unmarshal(respBytes, &responseData); err != nil {
+		return nil, []error{
+			fmt.Errorf(`Error fetching accounts %v via http: malformed response`, accountIDs),
```
Same as above. Consider adding the actual error to this log message

laurb9 replied:
Will replace with
```go
fmt.Errorf(`Error fetching accounts %v via http: failed to parse response: %v`, accountIDs, err),
```

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -81,8 +98,68 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
+			DataType: "Account",
+		}}
+	}
+	return accountJSON, []error{}
```
Please consider returning `nil` her for error rather than an empty slice as that's the preferred way in Go
Ref: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

laurb9 replied:
Will do, thanks for the ref link

mansinahar commented on stored_requests/backends/http_fetcher/fetcher_test.go:
```diff
@@ -61,14 +63,39 @@ func TestReqsAndImps(t *testing.T) {
 	assertErrLength(t, errs, 0)
 }
 
-func TestMissingValues(t *testing.T) {
```
Curious, what's the reason for removing this missing value test case?

laurb9 replied:
accident, will add back

mansinahar commented on stored_requests/backends/http_fetcher/fetcher_test.go:
```diff
@@ -61,14 +63,39 @@ func TestReqsAndImps(t *testing.T) {
+	fetcher, close := newFetcherBrokenBackend()
+	defer close()
+	accData, errs := fetcher.FetchAccounts(context.Background(), []string{"req-1"})
+	assertMapKeys(t, accData)
```
Please consider using `assert.Len` for this

laurb9 replied:
will replace with `assert.Empty`
will also replace `assertErrLength` with either `assert.Len` or `assert.Empty`, it does not appear to bring much to the table.

mansinahar commented on stored_requests/backends/http_fetcher/fetcher_test.go:
```diff
@@ -61,14 +63,39 @@ func TestReqsAndImps(t *testing.T) {
+
+	account, errs := fetcher.FetchAccount(context.Background(), "acc-1")
+	assert.Empty(t, errs, "Unexpected error fetching existing account")
+	assert.JSONEq(t, `"acc-1"`, string(account))
```
Consider adding a message to this assertion in case the assertion fails

mansinahar commented on stored_requests/backends/http_fetcher/fetcher_test.go:
```diff
@@ -61,14 +63,39 @@ func TestReqsAndImps(t *testing.T) {
+	assert.Empty(t, errs, "Unexpected error fetching existing account")
+	assert.JSONEq(t, `"acc-1"`, string(account))
+}
+
```
Please consider adding a test case for an invalid json response from the http endpoint.

Also, if possible consider updating the tests to a table test format as they are easier to follow and also easier to add/modify/update test cases.

laurb9 replied:
Added invalid json test.

laurb9 replied:
Regarding the table test: the tests are small and differ in almost every line; the table would have to include all the differences and the eventual custom text. It may not be much of a win considering the code that needs to be added to read and execute the tests will be more dense than any of them.

mansinahar replied:
Fair enough! I understand that it will be a lot more work to refactor these tests.

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -34,21 +39,33 @@ import (
+	// `&request-ids=...&imp-ids=...`.
+
+	if url, err := url.Parse(endpoint); err != nil {
+		glog.Errorf(`Invalid endpoint "%s": %v`, endpoint, err)
```
This should be `golg.Fatalf`. `Errorf` will just log the error and move on vs `Fatalf` will log the error and call `os.Exit(255)` and for an error like this, we wanna fail fast.

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -81,8 +98,68 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
+			fmt.Errorf(`Error fetching accounts %v via http: failed to parse response: %v`, accountIDs, err),
+		}
+	}
+	errs := convertNullsToErrs(responseData.Accounts, "Account", nil)
```
In this case I think it will be much more clearer to pass an empty error slice to the `convertNullsToErrs` function. In case of returns a `nil` value makes more sense.

mansinahar reviewed (commented):
LGTM in general. Just some minor comments and also, please consider adding assertion messages in the assert calls in `fetcher_test.go`

laurb9 commented:
added commit 1154006 to address feedback
- Bail out (Fatalf) if http endpoint isn't parseable.
- Add custom messages to test assertions.

Also moved error assertions before the data assertions to make it easier to understand test failures.

bsardo commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -34,21 +39,33 @@ import (
+
+	if url, err := url.Parse(endpoint); err != nil {
+		glog.Fatalf(`Invalid endpoint "%s": %v`, endpoint, err)
+	} else {
+		glog.Infof("Making http_fetcher for endpoint %v", url)
+	}
```
Nitpick: I don't think this needs to be in an `else` block now given the `Fatalf`.

laurb9 replied:
It was to keep url in scope but I can do without

bsardo commented on stored_requests/backends/http_fetcher/fetcher_test.go:
```diff
@@ -9,66 +9,113 @@ import (
+	assert.Len(t, errs, 3, "Fetching 3 unknown reqs+imps should return 3 errors")
+}
+
+func TestFetchAccounts(t *testing.T) {
```
The UT code coverage in this package is lacking but I think that's mostly attributed to the fact that `FetchCategories` doesn't have any tests. We'll backfill that some other time. With that said, I think we can rather easily get coverage for a lot of the `FetchAccount` and `FetchAccounts` error cases by adding a few additional tests:

```
func TestFetchAccountsNoIDsProvided(t *testing.T) {
	fetcher, close := newTestAccountFetcher(t, []string{"acc-1", "acc-2"})
	defer close()

	accData, errs := fetcher.FetchAccounts(nil, []string{})
	assert.Len(t, errs, 0)
	assert.Nil(t, accData)
}

// Force build request failure by not providing a context
func TestFetchAccountsFailedBuildRequest(t *testing.T) {
	fetcher, close := newTestAccountFetcher(t, []string{"acc-1", "acc-2"})
	defer close()

	accData, errs := fetcher.FetchAccounts(nil, []string{"acc-1"})
	assert.Len(t, errs, 1)
	assert.Nil(t, accData)
}

// Force http error via request timeout
func TestFetchAccountsContextTimeout(t *testing.T) {
	fetcher, close := newTestAccountFetcher(t, []string{"acc-1", "acc-2"})
	defer close()

	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(0))
	defer cancel()
	accData, errs := fetcher.FetchAccounts(ctx, []string{"acc-1"})
	assert.Len(t, errs, 1)
	assert.Nil(t, accData)
}

func TestFetchAccountNoIDProvided(t *testing.T) {
	fetcher, close := newTestAccountFetcher(t, nil)
	defer close()

	account, errs := fetcher.FetchAccount(context.Background(), "")
	assert.Len(t, errs, 1)
	assert.Nil(t, account)
}
```

laurb9 replied:
Will cut and paste :)
